### DISCUSSION TITLE
Enhancing the Phone Number Formatter for Ethiopia

### DISCUSSION THREAD

**Discussion by eyopa21** [HuluWZ/phone-formatter-eth#7] [Q&A] [answered]
### Introduction:
Welcome, developers! We're excited to introduce our npm package, **phone-formatter**, specifically designed for formatting and validating Ethiopian phone numbers. Whether you're working on a local project or an international application, this package aims to simplify handling Ethiopian numbers in different formats.

### Key Features:
- **Validation:** Ensure phone numbers follow the correct Ethiopian format, either as `+251XXXXXXXXX` or `09XXXXXXXX`.
- **Format Conversion:** Easily switch between international, local, and E.164 formats.
- **Normalization:** Remove spaces, dashes, or parentheses to maintain a clean, consistent format.
- **Carrier Detection:** Identify the telecom provider based on the phone number's prefix.

### Discussion Points:
1. **Usage Scenarios:** How have you integrated phone number formatting into your applications? What challenges have you faced with Ethiopian phone numbers?
2. **Feature Requests:** What additional features would you like to see in this package? (e.g., support for other regions, more formatting options, etc.)
3. **Community Contributions:** Interested in contributing? Let's discuss potential enhancements and how we can collaborate to improve this package further.
4. **Best Practices:** Share your best practices for handling phone numbers in applications, especially in terms of validation, formatting, and user experience.
5. **Real-World Use Cases:** If you've used this package in a project, we'd love to hear about your experience! How did it help streamline your development process?

### Call to Action:
Dive into the conversation! Whether you have questions, ideas, or feedback, this is the place to share and collaborate with others working on similar projects. Let's build a robust tool together that can benefit the entire community.

**Comment by HuluWZ** [comment, marked as answer]
Thank you for the feedback and interest in the **phone-formatter-eth** npm package! As the creator of this package, I’m excited to share how it’s been utilized and some technical insights:

### Usage Scenarios:
In my own projects, I've integrated **phone-formatter-eth** to manage Ethiopian phone numbers across various systems. For example, in a Node.js-based user registration service, I use the package to validate and format phone numbers before saving them to the database. Here’s a snippet of how I integrate it:

```javascript
const { formatPhone, isValid } = require('phone-formatter-eth');

app.post('/register', (req, res, next) => {
    const { phoneNumber } = req.body;
    const formattedNumber = formatPhone(phoneNumber);
    if (isValid(formattedNumber)) {
        req.body.phoneNumber = formattedNumber;
        next();
    } else {
        res.status(400).json({ error: 'Invalid phone number' });
    }
});
```
### Best Practices

When using **phone-formatter-eth**, I recommend combining it with additional validation tools to enhance robustness. For example, integrating it with **Joi** for server-side validation ensures that phone numbers are correctly formatted and validated before being processed.

### Example Integration with Joi

To integrate **phone-formatter-eth** with **Joi**, you can create a custom validation rule that formats and validates phone numbers. Here’s how you can set it up:

```javascript
const Joi = require('joi');
const { formatPhone, isValid } = require('phone-formatter-eth');

const schema = Joi.object({
    phoneNumber: Joi.string().custom((value, helpers) => {
        // Format the phone number to the international format
        const formatted = formatPhone(value');
        
        // Validate the formatted phone number
        if (!isValid(formatted)) {
            return helpers.error('Invalid phone number format');
        }
        
        // Return the formatted phone number if valid
        return formatted;
    })
});
```
